src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -0,0 +1,127 @@
+#include "ray/gcs/redis_gcs_client.h"
+
+namespace ray {
+namespace gcs {
+
+GcsNodeManager::GcsNodeManager(boost::asio::io_service &io_service,
+                               std::shared_ptr<gcs::RedisGcsClient> gcs_client)
+    : client_call_manager_(io_service),
+      gcs_client_(std::move(gcs_client)),
+      num_heartbeats_timeout_(RayConfig::instance().num_heartbeats_timeout()),
+      heartbeat_timer_(io_service) {
+  Start();
+}
+
+void GcsNodeManager::HandleHeartbeat(const ClientID &node_id,
+                                     rpc::HeartbeatTableData &&heartbeat_data) {
+  heartbeats_[node_id] = num_heartbeats_timeout_;
+  heartbeat_buffer_[node_id] = heartbeat_data;
+}
+
+void GcsNodeManager::Start() {
+  RAY_LOG(INFO) << "Starting gcs node manager.";
+  const auto lookup_callback = [this](Status status,
+                                      const std::vector<GcsNodeInfo> &node_info_list) {
+    for (const auto &node_info : node_info_list) {
+      if (node_info.state() != rpc::GcsNodeInfo::DEAD) {
+        // If there're any existing alive clients in client table, add them to
+        // our `heartbeats_` cache. Thus, if they died before monitor starts,
+        // we can also detect their death.
+        // Use `emplace` instead of `operator []` because we just want to add this
+        // client to `heartbeats_` only if it has not yet received heartbeat event.
+        // Besides, it is not necessary to add an empty `HeartbeatTableData`
+        // to `heartbeat_buffer_` as it doesn't make sense to broadcast an empty
+        // message to the cluster and it's ok to add it when actually receive
+        // its heartbeat event.
+        heartbeats_.emplace(ClientID::FromBinary(node_info.node_id()),
+                            num_heartbeats_timeout_);
+      }
+    }
+    Tick();
+  };
+  RAY_CHECK_OK(gcs_client_->Nodes().AsyncGetAll(lookup_callback));
+}
+
+/// A periodic timer that checks for timed out clients.
+void GcsNodeManager::Tick() {
+  DetectDeadNodes();
+  SendBatchedHeartbeat();
+  ScheduleTick();
+}
+
+void GcsNodeManager::DetectDeadNodes() {
+  for (auto it = heartbeats_.begin(); it != heartbeats_.end();) {
+    it->second = it->second - 1;
+    if (it->second == 0) {
+      if (dead_nodes_.count(it->first) == 0) {
+        auto node_id = it->first;
+        RAY_LOG(WARNING) << "Node timed out: " << node_id;
+        auto lookup_callback = [this, node_id](Status status,
+                                               const std::vector<GcsNodeInfo> &all_node) {
+          RAY_CHECK_OK(status);
+          bool marked = false;
+          for (const auto &node : all_node) {
+            if (node_id.Binary() == node.node_id() && node.state() == GcsNodeInfo::DEAD) {
+              // The node has been marked dead by itself.
+              marked = true;
+              break;
+            }
+          }
+          if (!marked) {
+            RAY_CHECK_OK(gcs_client_->Nodes().AsyncUnregister(node_id, nullptr));
```

Comment:
it's a bit strange to have `gcs_client_`  here inside gcs server code. Maybe can rename later when we have separated pubsub and backend storage.

Reply:
Sounds greet, but @ffbin is working on the interface of backend storage and after then the `gcs_client_` will be replaced by `storage_client_`

Reply:
> it's a bit strange to have `gcs_client_` here inside gcs server code. Maybe can rename later when we have separated pubsub and backend storage.

Sounds greet, but @ffbin is working on the interface of backend storage and after then the `gcs_client_` will be replaced by `storage_client_`.